src/fal/dag.py

```diff
@@ -0,0 +1,129 @@
+from dataclasses import dataclass, field
+from os import path
+from typing import List,TypeVar, Dict
+from faldbt.project import DbtModel 
+from pathlib import Path
+import re
+
+T = TypeVar("T", bound="FalScript")
+
+@dataclass(frozen=True)
+class FalScript:
+    model: DbtModel
+    path: str
+    dependencies: List[T] = field(default_factory=list)
+
+    def __hash__(self):
+        return self.path.__hash__()
+
+    def __repr__(self):
+        return "FalScript(" + self.path.__str__() + ")"
+
+    def parse(self) -> List[str]:
+        """
+        Parses the python file to identify any dependencies
+        """
+        with open(self.path) as file:
+            lines = file.readlines()
+            ## TODO: do this search in parallel
+            dependency_names = []
+            for line in lines:
+                match = re.search('ref\("(.*)"\)', line)
+                if match:
+                    group = match.group(1)
+                    dependency_names.append(group)
+            return dependency_names    
+    
+    def addDependency(self, dependency: T):
+        if (dependency not in self.dependencies):
+            self.dependencies.append(dependency)
+
+    def exec(self, ref, context, source):
+        """
+        Executes the script
+        """
+        with open(path) as file:
+            a_script = file.read()
+            exec(
+                a_script,
+                {
+                    "ref": ref,
+                    "context": context,
+                    "source": source,
+                },
+            )
+
+@dataclass(frozen=True)
+class UniqueKey:
+    model: DbtModel
+    script_path: Path
+
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
+    falScripts: Dict[UniqueKey, FalScript] = {}
+
+    def __init__(self, models: List[DbtModel], keyword: str, root: str):
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursivelySetDependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursivelySetDependencies(self, key: UniqueKey):
+        if (self.falScripts.get(key) is None):
+            self.falScripts[key] = FalScript(key.model, key.script_path)
```

Comment:
What if we change this first few lines to check if it is already set and return right away? I think once it is set, we don't want to continue with the rest of the script, right?

```py
if key in self.falScripts:
  return

current_script = FalScript(...)
self.falScripts[key] = current_script
```

The interesting thing about that is that then inside the `for`, in line 87 we no longer have to check before calling. We can call `self.recursivelySetDependencies(UniqueKey(model, dependency_scipt))` and it checks for us in the top of the method.